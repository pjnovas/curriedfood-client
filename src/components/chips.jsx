import React from 'react';
import PropTypes from 'prop-types';
import { ScrollView, StyleSheet } from 'react-native';
import { Chip } from 'react-native-paper';

import { noop, xor } from 'lodash';
import { composeHooks } from 'utils/language';

export const Chips = ({ list, items, onPress, ...props }) => (
  <ScrollView>
    {list.map((name) => (
      <Chip
        key={name}
        selected={items.includes(name)}
        mode="outlined"
        style={styles.chip}
        onPress={onPress(name)}
        {...props}
      >
        {name}
      </Chip>
    ))}
  </ScrollView>
);

Chips.propTypes = {
  ...Chip.PropTypes,
  list: PropTypes.arrayOf(PropTypes.string).isRequired,
  items: PropTypes.arrayOf(PropTypes.string),
  onPress: PropTypes.func
};

Chips.defaultProps = {
  items: [],
  onPress: noop
};

const styles = StyleSheet.create({
  chip: {
    marginVertical: 5,
    borderRadius: 4
  }
});

export const useFilter = ({ value, onChange }) => {
  const items = value ? value.split(',') : [];

  return {
    items,
    onPress: (name) => () => onChange(xor(items, [name]).join(','))
  };
};

export default composeHooks({ useFilter })(Chips);
