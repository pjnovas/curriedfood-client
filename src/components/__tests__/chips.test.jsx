import { noop } from 'lodash';
import { Chips, useFilter } from '../chips';

const list = ['almacen', 'super', 'verduleria', 'carniceria'];

const props = {
  list,
  onPress: noop
};

takeSnapshots(
  {
    default: props,
    withItems: {
      ...props,
      items: ['verduleria', 'super']
    }
  },
  Chips
);

describe('useFilter', () => {
  const value = 'verduleria,super,almacen';
  const onChange = jest.fn();

  beforeEach(() => {
    onChange.mockClear();
  });

  test('translates a coma separated string into an array of items', () => {
    expect(useFilter({ value, onChange }).items).toEqual([
      'verduleria',
      'super',
      'almacen'
    ]);

    expect(useFilter({ onChange }).items).toEqual([]);
  });

  test('parses the list when an item is pressed', () => {
    const result = useFilter({ value, onChange });

    expect(result.items).toEqual(['verduleria', 'super', 'almacen']);

    result.onPress('super')(); // exists: removes it
    expect(onChange).toBeCalledWith('verduleria,almacen');

    result.onPress('carniceria')(); // does not exists: add it
    expect(onChange).toBeCalledWith('verduleria,super,almacen,carniceria');
  });
});
