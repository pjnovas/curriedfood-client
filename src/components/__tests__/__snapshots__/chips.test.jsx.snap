// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default 1`] = `
<ScrollViewMock>
  <withTheme(Chip)
    key="almacen"
    mode="outlined"
    selected={false}
    style={
      Object {
        "borderRadius": 4,
        "marginVertical": 5,
      }
    }
  >
    almacen
  </withTheme(Chip)>
  <withTheme(Chip)
    key="super"
    mode="outlined"
    selected={false}
    style={
      Object {
        "borderRadius": 4,
        "marginVertical": 5,
      }
    }
  >
    super
  </withTheme(Chip)>
  <withTheme(Chip)
    key="verduleria"
    mode="outlined"
    selected={false}
    style={
      Object {
        "borderRadius": 4,
        "marginVertical": 5,
      }
    }
  >
    verduleria
  </withTheme(Chip)>
  <withTheme(Chip)
    key="carniceria"
    mode="outlined"
    selected={false}
    style={
      Object {
        "borderRadius": 4,
        "marginVertical": 5,
      }
    }
  >
    carniceria
  </withTheme(Chip)>
</ScrollViewMock>
`;

exports[`withItems 1`] = `
<ScrollViewMock>
  <withTheme(Chip)
    key="almacen"
    mode="outlined"
    selected={false}
    style={
      Object {
        "borderRadius": 4,
        "marginVertical": 5,
      }
    }
  >
    almacen
  </withTheme(Chip)>
  <withTheme(Chip)
    key="super"
    mode="outlined"
    selected={true}
    style={
      Object {
        "borderRadius": 4,
        "marginVertical": 5,
      }
    }
  >
    super
  </withTheme(Chip)>
  <withTheme(Chip)
    key="verduleria"
    mode="outlined"
    selected={true}
    style={
      Object {
        "borderRadius": 4,
        "marginVertical": 5,
      }
    }
  >
    verduleria
  </withTheme(Chip)>
  <withTheme(Chip)
    key="carniceria"
    mode="outlined"
    selected={false}
    style={
      Object {
        "borderRadius": 4,
        "marginVertical": 5,
      }
    }
  >
    carniceria
  </withTheme(Chip)>
</ScrollViewMock>
`;
